Type narrowing not working in certain cases

### Vue - Official extension or vue-tsc version

3.0.0-beta.2

### VSCode version

\-

### Vue version

3.5.16

### TypeScript version

5.8.3

### System Info

stackblitz

### package.json dependencies

```json
{
  "type": "module",
  "devDependencies": {
    "@vue/tsconfig": "^0.7.0",
    "typescript": "^5.8.3",
    "vue": "^3.5.16",
    "vue-tsc": "^3.0.0-beta.2"
  }
}
```

### Steps to reproduce

See link below. Run `vue-tsc --noEmit`.

```vue
<script setup lang="ts">
import { computed } from 'vue';
import AButton from './AButton.vue';

const file = computed<{
  action?: { onClick(): void };
}>(() => ({}));
</script>

<template>
  <div v-if="file.action">
    <AButton @click="file.action.onClick">error</AButton>
    <button @click="file.action.onClick">no error</button>
  </div>
</template>
```

### What is expected?

No error in both cases.

### What is actually happening?

```vue
Foo.vue:12:22 - error TS18048: '__VLS_ctx.file.action' is possibly 'undefined'.

12     <AButton @click="file.action.onClick">error</AButton>
                        ~~~~~~~~~~~


Found 1 error in Foo.vue:12
```

### Link to minimal reproduction

https://stackblitz.com/edit/node-pfaigjdj?file=Foo.vue

### Any additional comments?

Working fine on v2.

--- a/packages/language-core/lib/codegen/template/elementEvents.ts
+++ b/packages/language-core/lib/codegen/template/elementEvents.ts
@@ -58,7 +58,7 @@ export function* generateElementEvents(
 			const emitName = emitPrefix + source;
 			const camelizedEmitName = camelize(emitName);
 
-			yield `(): __VLS_NormalizeComponentEvent<typeof ${propsVar}, typeof ${emitsVar}, '${propName}', '${emitName}', '${camelizedEmitName}'> => (${newLine}`;
+			yield `const ${ctx.getInternalVariable()}: __VLS_NormalizeComponentEvent<typeof ${propsVar}, typeof ${emitsVar}, '${propName}', '${emitName}', '${camelizedEmitName}'> = (${newLine}`;
 			if (prop.name === 'on') {
 				yield `{ `;
 				yield* generateEventArg(ctx, source, start!, emitPrefix.slice(0, -1), ctx.codeFeatures.navigation);
@@ -121,10 +121,12 @@ export function* generateEventExpression(
 		let isFirstMapping = true;
 
 		const ast = createTsAst(options.ts, prop.exp, prop.exp.content);
-		const _isCompoundExpression = isCompoundExpression(options.ts, ast);
-		if (_isCompoundExpression) {
-			yield `(...[$event]) => {${newLine}`;
+		const isCompound = isCompoundExpression(options.ts, ast);
+
+		if (isCompound) {
 			ctx.addLocalVariable('$event');
+
+			yield `(...[$event]) => {${newLine}`;
 			yield* ctx.generateConditionGuards();
 			prefix = ``;
 			suffix = ``;
@@ -135,7 +137,7 @@ export function* generateEventExpression(
 			ctx,
 			'template',
 			offset => {
-				if (_isCompoundExpression && isFirstMapping) {
+				if (isCompound && isFirstMapping) {
 					isFirstMapping = false;
 					ctx.inlayHints.push({
 						blockName: 'template',
@@ -159,7 +161,7 @@ export function* generateEventExpression(
 			suffix
 		);
 
-		if (_isCompoundExpression) {
+		if (isCompound) {
 			ctx.removeLocalVariable('$event');
 
 			yield endOfLine;

--- a/test-workspace/tsc/passedFixtures/vue3/#5428/main.vue
+++ b/test-workspace/tsc/passedFixtures/vue3/#5428/main.vue
@@ -1,0 +1,14 @@
+<script setup lang="ts">
+let foo!: {
+	bar?: {
+		baz: () => void;
+	}
+};
+const Comp = () => {};
+</script>
+
+<template>
+	<div v-if="foo.bar">
+		<Comp @click="foo.bar.baz" />
+	</div>
+</template>

